uaclient/sso.py

```diff
@@ -173,10 +174,14 @@ def prompt_request_macaroon(cfg: UAConfig, caveat_id: str) -> dict:
         try:
             content = sso_client.request_discharge_macaroon(**args)
         except SSOAuthError as e:
-            if API_ERROR_2FA_REQUIRED not in e:
-                raise exceptions.UserFacingError(str(e))
-            args['otp'] = input('Second-factor auth: ')
-            continue
+            if API_ERROR_2FA_REQUIRED in e:
+                args['otp'] = input('Second-factor auth: ')
+                continue
+            elif API_ERROR_INVALID_CREDENTIALS in e:
+                if '2-factor' in e[API_ERROR_INVALID_CREDENTIALS]:
```

Comment:
Is this comparing against the human-readable message provided by the contract server? Are we sure that message is considered stable by the contract server team?

Reply:
It's comparing thea human-readable message for login.ubuntu.com which is publicly documented as error values. 
https://canonical-identity-provider.readthedocs.io/en/latest/resources/token.html?highlight=invalid-credentials


Just to be sure I confirmed with roadmr in #snapstore channel today that those are expected stable values and he said there are no plans in the future  to change value strings and that part of the code hasn't changed in years.

Reply:
That documentation appears to suggest that we should be seeing `TWOFACTOR_FAILURE` as the error, not `INVALID_CREDENTIALS`; am I missing something?

Reply:
@OddBloke I checked with roadmr about it. He says while it may be an ommision/documentation error, they will not plan to change the error code as they might break existing clients. For the same reason they will also not change the error message because clients may currently be parsing it.

Reply:
This makes me nervous, because we have to hope that this decision (which is some combination of undocumented and incorrectly documented) doesn't change in the ~8.5 years for which we will be supporting bionic images with the client code that relies on this behaviour.

That said, it's not like we really have a choice of upstream here so I don't see what other option we have, but I'd like @powersj to be aware of this, and sign-off before we land this.

Reply:
Sounds fair @OddBloke, though we could future-proof ourselves in the event that they did add an error 'code': 'twofactor-failure' in the future as properly documented.

Reply:
Two changes:

1) Only allow the user to re-try 2fa once.
2) I want a comment explainging the use of `INVALID_CREDENTIALS` over `TWOFACTOR_FAILURE`

Reply:
Done and done. Thanks gents.